### PR TITLE
Move ingest processors from core to ingest-common module.

### LINKED ISSUE #18490: Move all ingest processors in core to a module?
This module should only contain the common processors that have external dependencies or require additional security permissions.

Not sure what the name should be. Maybe `ingest-common` or `ingest-processors`.

The main benefit would be that during development of common ingest processors, running tests becomes much quicker. Like @nik9000 mentioned [here](https://github.com/elastic/elasticsearch/issues/17317#issuecomment-200809490) as a good reason to move the percolator to a module.

### PR DESCRIPTION
**Pull request by martijnvg**
Also all the rest tests have been moved to ingest-common module as well (rest tests can't be ran without processor implementations), because these tests don't run in the rest-api-spec module but in the distribution:integ-test-zip module and adding a test plugin with a mock plugin there felt just wrong to me. I think this is fine. I left a tiny ingest rest test behind in that tests with an empty pipeline.

Closes #18490

### CODE REVIEW

**Comment by martijnvg** [comment]
Made a couple of updates:
* Moved reindex ingest tests to `smoke-test-ingest-with-all-dependencies` qa module.
* Extracted test plugin from core to test infra, so that the reindex module can reuse that.
* Removed two messy tests from `smoke-test-ingest-with-all-dependencies` qa module. What these tests tested was already covered by the rest tests in this qa module.

File: qa/smoke-test-ingest-with-all-dependencies/src/test/java/org/elasticsearch/ingest/IngestMustacheSetProcessorIT.java [hunk 1]
```diff
@@ -1,72 +0,0 @@
-/*
```
Comment (talevy) [inline qa/smoke-test-ingest-with-all-dependencies/src/test/java/org/elasticsearch/ingest/IngestMustacheSetProcessorIT.java]: why was this removed?

**Comment by talevy** [comment]
Looks good. will make testing way nicer!

File: qa/smoke-test-ingest-with-all-dependencies/src/test/java/org/elasticsearch/ingest/IngestMustacheSetProcessorIT.java (hunk 1 above)
Comment (martijnvg) [inline qa/smoke-test-ingest-with-all-dependencies/src/test/java/org/elasticsearch/ingest/IngestMustacheSetProcessorIT.java]: Because we test this already in the rest tests that are part of this module. If templating is broken then these rest tests will fail too, so it safe to remove these tests.

**Comment by martijnvg** [comment]
@talevy I've also moved the grok processor to the ingest-commons module and dropped the ingest-grok module like we discussed yesterday.

File: modules/ingest-common/src/main/java/org/elasticsearch/ingest/IngestCommonsPlugin.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.nio.charset.StandardCharsets;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+public class IngestCommonsPlugin extends Plugin {
```
Comment (rjernst) [inline modules/ingest-common/src/main/java/org/elasticsearch/ingest/IngestCommonsPlugin.java]: You have "commons" (plural) here, but the plugin name as "ingest-common". I would think singular is better, if only to not confuse with apache commons.

**Comment by rjernst** [comment]
I like the idea of moving this to a module, but why collapse in the grok module?

**Comment by javanna** [comment]
Not that it should block this PR and sorry if I am late to the party... I am genuinely afraid that we will end up with too many modules just because they make it easy to run a subset of the tests. I don't find this reason compelling enough and I wonder if there isn't a better solution to make development easier. Am I the only one having this concern? or maybe there are other reasons to have this as a module other than executing specific tests? I feel like the reasons why something should become a module are shifting from the initial "to isolate dependencies and crazy security permissions" to something else where many more things should become modules. Do modules have a cost in our build, or can we just have as many as we want? Maybe I am simply worrying too much though :)

**Comment by rjernst** [comment]
> Do modules have a cost in our build, or can we just have as many as we want?

The only cost is a slightly longer configuration time in gradle, but this continues to improve with new versions of gradle.

>  I feel like the reasons why something should become a module are shifting from the initial "to isolate dependencies and crazy security permissions" to something else where many more things should become modules

To me, in addition to dependencies and security, it has to do with "keeping core small": #10368. It also has the added benefit that it forces us to ensure these things are actually pluggable, and the tests in core can be very generic for that exact purpose (does the infrastructure work, vs does a specific implementation work).

**Comment by talevy** [comment]
Regarding the easy testing comment, maybe this warrants a separate issue?

This is how I currently test ingest changes

```
gradle :core:check -Dtests.class=org.elasticsearch.action.ingest.*
gradle :core:check -Dtests.class=org.elasticsearch.ingest.*
gradle :modules:ingest-grok:check
gradle :plugins:ingest-geoip:check
gradle :qa:smoke-test-ingest-with-all-dependencies:check
gradle :distribution:integ-test-zip:integTest -Dtests.class=org.elasticsearch.test.rest.RestIT
```

these don't take too long and they avoid running tests for all of core.

**Comment by javanna** [comment]
I like the "keep core small" and "verify pluggability" reasons much more than "make testing faster" alone. I am convinced this makes sense, thanks @rjernst for your thoughts.

File: modules/ingest-common/src/main/java/org/elasticsearch/ingest/IngestCommonsPlugin.java (hunk 2 above)
Comment (martijnvg) [inline modules/ingest-common/src/main/java/org/elasticsearch/ingest/IngestCommonsPlugin.java]: good point. I'll rename the module.

**Comment by martijnvg** [comment]
> I am genuinely afraid that we will end up with too many modules just because they make it easy to run a subset of the tests. 

@javanna Agreed, we should end up with too many modules (like a module per processor).  It should be balanced and I think we have the right balance here by adding all processors that don't require additional security permissions (grok, set, date, etc) into a single module. This way we benefit from a lean and mean ES core and testing ingest processors being quick.

> we will still have to run core tests since the core ingest apis and Pipeline features are still in core.

Yes, changes to the ingest infra in core require testing several modules. However changing just a processor will require to only test a single plugin or module and the `smoke-test-ingest-with-all-dependencies` qa module.

File: modules/ingest-common/src/main/java/org/elasticsearch/ingest/IngestCommonsPlugin.java (hunk 2 above)
Comment (martijnvg) [inline modules/ingest-common/src/main/java/org/elasticsearch/ingest/IngestCommonsPlugin.java]: Turns this is just a inconsistency, the module itself has the right name...

File: modules/ingest-common/build.gradle [hunk 3]
```diff
@@ -18,8 +18,8 @@
  */
 
 esplugin {
-    description 'Ingest processor that uses grok patterns to split text'
-    classname 'org.elasticsearch.ingest.grok.IngestGrokPlugin'
+    description 'Module for ingest processors that do not require additional security permissions.'
```
Comment (talevy) [inline modules/ingest-common/build.gradle]: > do not require additional security permissions.

is that the actually what *will* be shared by processors that can be found in this module? Isn't the reason we moved geoip into its own plugin because of package size, not because of its security permissions?

File: modules/ingest-common/src/main/java/org/elasticsearch/ingest/IngestCommonPlugin.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    private static final String[] PATTERN_NAMES = new String[] {
+            "aws", "bacula", "bro", "exim", "firewalls", "grok-patterns", "haproxy",
+            "java", "junos", "linux-syslog", "mcollective-patterns", "mongodb", "nagios",
+            "postgresql", "rails", "redis", "ruby"
+    };
+
+    private final Map<String, String> builtinPatterns;
+
+    public IngestCommonPlugin() throws IOException {
+        this.builtinPatterns = loadBuiltinPatterns();
```
Comment (talevy) [inline modules/ingest-common/src/main/java/org/elasticsearch/ingest/IngestCommonPlugin.java]: should we add comments around this grok-specific setup that can now be found in the general IngestCommonPlugin? just in case this gets factored out in the future

File: modules/ingest-common/build.gradle (hunk 3 above)
Comment (martijnvg) [inline modules/ingest-common/build.gradle]: Good point. I'll adjust the description, so that it also describes that we don't want processors with large resources / dependencies in this module.

File: modules/ingest-common/src/main/java/org/elasticsearch/ingest/IngestCommonPlugin.java (hunk 4 above)
Comment (martijnvg) [inline modules/ingest-common/src/main/java/org/elasticsearch/ingest/IngestCommonPlugin.java]: +1 I'll add a comment

**Comment by dadoonet** [comment]
So apparently modules are not published to maven central?

Here is my use case. I'm building an ingest plugin for elasticsearch using Maven.

I depend on:

```xml
        <dependency>
            <groupId>org.elasticsearch</groupId>
            <artifactId>elasticsearch</artifactId>
            <version>5.0.0-alpha4</version>
            <scope>provided</scope>
        </dependency>
```

Then, I want to run a simple test where I'm building a pipeline aggregation using:

* rename processor
* my own processor

Was working well with alpha3 because rename was in core. Now it moved to a module which is not available on maven central so I can't use the code in my tests.
I'm now getting `No processor type exists with name [rename]"`.

This is a similar concern as I explained at #18131.

Any chance we will publish `ingest-common` JAR on maven central?

**Comment by martijnvg** [comment]
> So apparently modules are not published to maven central?

Correct, modules aren't published. I think the reason is because they are considered internal to ES, but @rjernst can confirm this.

Ideally any of your (unit and integration) tests shouldn't depend on other processors. If you do want to test an use case that requires multiple processors, then it is better to add such a test to a qa module. If your processor test depends on some really basic processor operation, then you can also use a mock test processor in your test (See TestProcessor in test framework).
